clientv3/watch.go

```diff
@@ -456,11 +462,23 @@ func (w *watchGrpcStream) run() {
 					wc.Send(ws.initReq.toPB())
 				}
 			case pbresp.Canceled:
-				delete(cancelSet, pbresp.WatchId)
-				if ws, ok := w.substreams[pbresp.WatchId]; ok {
-					// signal to stream goroutine to update closingc
-					close(ws.recvc)
-					closing[ws] = struct{}{}
+				if pbresp.CancelReason == pb.WatchResponse_PERMISSION_DENIED {
```

Comment:
I thought the special handling is required because the error of permission denied will be replied from etcd to the client before registering a substream so the new `dispatchCanceledEvent()` is needed. Otherwise, the client cannot know the reason of the cancel. In a case of etcdctl, the command cannot print a detailed reason for a user.